assets/js/blocks/product-collection/inspector-controls/inherit-query-control.tsx

```diff
@@ -0,0 +1,112 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { select } from '@wordpress/data';
+import { isSiteEditorPage } from '@woocommerce/utils';
+import { usePrevious } from '@woocommerce/base-hooks';
+import {
+	ToggleControl,
+	// @ts-expect-error Using experimental features
+	// eslint-disable-next-line @wordpress/no-unsafe-wp-apis
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { ProductCollectionQuery } from '../types';
+import { DEFAULT_QUERY } from '../constants';
+
+const ARCHIVE_PRODUCT_TEMPLATES = [
+	'woocommerce/woocommerce//archive-product',
+	'woocommerce/woocommerce//taxonomy-product_cat',
+	'woocommerce/woocommerce//taxonomy-product_tag',
+	'woocommerce/woocommerce//taxonomy-product_attribute',
+	'woocommerce/woocommerce//product-search-results',
+];
+
+const label = __(
+	'Inherit query from template',
+	'woo-gutenberg-products-block'
+);
+
+interface InheritQueryControlProps {
+	setQueryAttribute: ( value: Partial< ProductCollectionQuery > ) => void;
+	query: ProductCollectionQuery | undefined;
+}
+
+const InheritQueryControl = ( {
+	setQueryAttribute,
+	query,
+}: InheritQueryControlProps ) => {
+	const inherit = query?.inherit;
+	const editSiteStore = select( 'core/edit-site' );
+	const currentTemplateId = editSiteStore?.getEditedPostId() as string;
+
+	const queryObjectBeforeInheritEnabled = usePrevious(
+		query,
+		( value?: ProductCollectionQuery ) => {
+			return value?.inherit === false;
+		}
+	);
+
+	/**
+	 * Set inherit value when Product Collection block is first added to the page.
+	 * We want inherit value to be true when block is added to ARCHIVE_PRODUCT_TEMPLATES
+	 * and false when added to somewhere else.
+	 */
+	const initialValue = currentTemplateId
+		? ARCHIVE_PRODUCT_TEMPLATES.includes( currentTemplateId )
+		: false;
+	if ( inherit === null ) {
+		setQueryAttribute( {
+			inherit: initialValue,
+		} );
```

Comment:
Sometimes, when adding the Product Collection block into the Site Editor, I get this JS warning in the console:

> Warning: Cannot update a component (`DocumentActions`) while rendering a different component (`InheritQueryControl`). To locate the bad setState() call inside `InheritQueryControl`, follow the stack trace as described in...

Can you reproduce as well?

I think the error is caused by this `setQueryAttribute()`. If I'm understanding the code correctly, it's trying to update the state during the render function, which causes an unnecessary re-render and might cause other side effects. I think we should set the value of `inherit` to `initialValue` right when we read it, so in `assets/js/blocks/product-collection/inspector-controls/index.tsx`. Would that work? :thinking:

Reply:
I can also see this warning. I have relocated the code responsible for calculating the inherit value to:
[edit.tsx](https://github.com/woocommerce/woocommerce-blocks/blob/34762c5190f87382370bcec12177ae0416f44299/assets/js/blocks/product-collection/edit.tsx#L96-L119).

Now, I am setting the inherit value while defining the default attributes. Additionally, I have made some other minor improvements in [34762c5](https://github.com/woocommerce/woocommerce-blocks/pull/9485/commits/34762c5190f87382370bcec12177ae0416f44299).

What are your thoughts on the recent changes? Do they look good to you?

Reply:
Yup, looks good. I initially thought the `useEffect()` would not be necessary, but I guess because of https://github.com/WordPress/gutenberg/issues/7342, there is no way we can avoid it. :disappointed: